### PR TITLE
Better failure recovery for build jobs

### PR DESCRIPTION
**Pull request by chefsalim**
This change adds additional error handling to the builder-jobsrv. If the worker heartbeats stop during the execution of an active job, the job is reset back to a Pending state (on the assumption that the previously assigned worker is dead), and re-assigned to a different worker.  Another case that is handled is a timeout for a job that is assigned to a worker - if the job is not completed within a set limit (currently, 90 mins), the job is assumed to be in a bad state, and is Failed.

As part of this change, the heartbeat protocol was modified to send heartbeats even when the worker is busy. Previously, when a worker was executing a build job, the heartbeats would be paused.

Signed-off-by: Salim Alam <salam@chef.io>

tenor-71988318

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-jobsrv/src/server/worker_manager.rs [hunk 1]
```diff
@@ -235,37 +249,118 @@ impl WorkerMgr {
...
                 if expiry >= &now {
                     break;
                 }
             } else {
                 break;
             }
-            let worker = self.workers.pop_front();
-            debug!("expiring worker due to inactivity, worker={:?}", worker);
+
+            let (worker, _) = self.ready_workers.pop_front().unwrap();
```
Comment (raskchanky) [inline components/builder-jobsrv/src/server/worker_manager.rs]: If `self.ready_workers` is empty, will this panic?

File: components/builder-jobsrv/src/server/worker_manager.rs [hunk 2]
```diff
@@ -235,37 +249,118 @@ impl WorkerMgr {
...
+        loop {
+            if let Some((_, expiry)) = self.busy_workers.front() {
+                if expiry >= &now {
+                    break;
+                }
+            } else {
+                break;
+            }
+
+            let (worker, _) = self.busy_workers.pop_front().unwrap();
```
Comment (raskchanky) [inline components/builder-jobsrv/src/server/worker_manager.rs]: Same comment here as above.  Do we need to be concerned about `busy_workers` being empty and this panicking?

File: components/builder-jobsrv/src/server/worker_manager.rs [hunk 3]
```diff
@@ -235,37 +249,118 @@ impl WorkerMgr {
...
+            debug!(
+                "expiring busy worker due to missed heartbeat, worker={:?}",
+                worker
+            );
+
+            assert!(self.worker_map.contains_key(&worker));
```
Comment (raskchanky) [inline components/builder-jobsrv/src/server/worker_manager.rs]: I might be overly paranoid, but an `assert!` will panic the current thread if it fails right?  Is that what we want here?

File: components/builder-jobsrv/src/server/worker_manager.rs [hunk 4]
```diff
@@ -235,37 +249,118 @@ impl WorkerMgr {
...
+        loop {
+            if let Some((_, expiry)) = self.jobs.front() {
+                if expiry >= &now {
+                    break;
+                }
+            } else {
+                break;
+            }
+
+            let (job_id, _) = self.jobs.pop_front().unwrap();
```
Comment (raskchanky) [inline components/builder-jobsrv/src/server/worker_manager.rs]: Same comment here as above, re: unwrap.

File: components/builder-jobsrv/src/server/worker_manager.rs (hunk 1 above)
Comment (chefsalim) [inline components/builder-jobsrv/src/server/worker_manager.rs]: No, we check the existence of an available entry in the if statement above and break if it is empty.

File: components/builder-jobsrv/src/server/worker_manager.rs (hunk 2 above)
Comment (chefsalim) [inline components/builder-jobsrv/src/server/worker_manager.rs]: No, same as above

File: components/builder-jobsrv/src/server/worker_manager.rs (hunk 3 above)
Comment (chefsalim) [inline components/builder-jobsrv/src/server/worker_manager.rs]: Yeah I went back and forth on that - for now, I'd rather have it panic if this logic is wrong.

File: components/builder-jobsrv/src/server/worker_manager.rs (hunk 4 above)
Comment (chefsalim) [inline components/builder-jobsrv/src/server/worker_manager.rs]: Same as above

File: components/builder-jobsrv/src/server/worker_manager.rs (hunk 1 above)
Comment (raskchanky) [inline components/builder-jobsrv/src/server/worker_manager.rs]: Ah, right.  I missed that 😞

**Comment by raskchanky** [comment]
tenor-37528076

**Comment by reset** [comment]
@thesentinels approve

tenor-94394616

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/258727322).

**Comment by thesentinels** [comment]
:broken_heart: Travis CI reports [this PR failed to pass the test suite](https://travis-ci.org/habitat-sh/habitat/builds/258727322).

The next step is to examine the [job](https://travis-ci.org/habitat-sh/habitat/builds/258727322) and figure out why. If it is transient, you can try re-triggering the Travis CI Job - if it passes, this PR will be automatically merged. If it is not transient, you should fix the issue and update this pull request, and issue `approve` again. If you believe it will never pass, and you are feeling :godmode:, you can issue a `force` to merge this PR anyway.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/258727322).

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/258727322).

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/258727322).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/258727322).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**